File: src/types/menu.ts
export type MenuSite = 'APPLICANT' | 'ADMIN'
export type MenuType = 'ROUTE' | 'URL'

export interface MenuItem {
  id: number
  parentId: number | null
  site: MenuSite
  type: MenuType
  name: string
  path: string | null
  sortOrder: number
  /*
   * ant-design-vue 아이콘 컴포넌트명을 문자열 그대로 담는다(예: 'SettingOutlined').
   * 관리자 좌측 사이드바 전용이며 지원자(가로 헤더) 메뉴는 null이다.
   */
  icon: string | null
  children: MenuItem[]
}

File: recruit_back/recruit_backend/src/main/java/com/shinyoung/recruit/domain/entity/Menu.java
package com.shinyoung.recruit.domain.entity;

import com.shinyoung.recruit.enumeration.MenuSite;
import com.shinyoung.recruit.enumeration.MenuType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Getter
@NoArgsConstructor
public class Menu extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Enumerated(EnumType.STRING)
    private MenuSite site;
    @Enumerated(EnumType.STRING)
    private MenuType type;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Menu parent;

    private String name;
    private String path;

    private Integer sortOrder;

    @Column(length = 100)
    private String icon;

    private Menu(
            MenuSite site,
            MenuType type,
            Menu parent,
            String name,
            String path,
            Integer sortOrder,
            String icon
    ) {
        this.site = site;
        this.type = type;
        this.parent = parent;
        this.name = name;
        this.path = path;
        this.sortOrder = sortOrder;
        this.icon = icon;
    }

    public static Menu create(
            MenuSite site,
            MenuType type,
            Menu parent,
            String name,
            String path,
            Integer sortOrder,
            String icon
    ) {
        return new Menu(site, type, parent, name, path, sortOrder, icon);
    }

    public void update(
            MenuSite site,
            MenuType type,
            Menu parent,
            String name,
            String path,
            Integer sortOrder,
            String icon
    ) {
        this.site = site;
        this.type = type;
        this.parent = parent;
        this.name = name;
        this.path = path;
        this.sortOrder = sortOrder == null ? 0 : sortOrder;
        this.icon = icon;
    }

    public boolean isRootMenu() {
        return this.parent == null;
    }

    public boolean isSubMenu() {
        return this.parent != null;
    }
}
